cpp/include/cudf/detail/gather.cuh

```diff
@@ -0,0 +1,256 @@
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+#include <cudf/copying.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cudf_utils.h>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <algorithm>
+
+#include <thrust/functional.h>
+#include <thrust/logical.h>
+#include <thrust/gather.h>
+#include <thrust/iterator/transform_iterator.h>
+
+
+#include <cub/cub.cuh>
+#include <utilities/column_utils.hpp>
+#include <utilities/cuda_utils.hpp>
+#include <utilities/release_assert.cuh>
+
+#include <bitmask/legacy/valid_if.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/**---------------------------------------------------------------------------*
+ * @brief Function object to check if an index is within the bounds [begin,
+ * end).
+ *
+ *---------------------------------------------------------------------------**/
+template <typename map_type>
+struct bounds_checker {
+  size_type begin;
+  size_type end;
+
+  __device__ bounds_checker(size_type begin_, size_type end_)
+    : begin{begin_}, end{end_} {}
+
+  __device__ bool operator()(map_type const index) {
+    return ((index >= begin) && (index < end));
+  }
+};
+
+
+template <bool ignore_out_of_bounds, typename map_iterator_type>
+__global__ void gather_bitmask_kernel(table_device_view source_table,
+				      map_iterator_type gather_map,
+				      mutable_table_device_view destination_table) {
+  for (size_type i = 0; i < source_table.num_columns(); i++) {
+
+    const int warp_size = 32;
+
+    column_device_view source_col = source_table.column(i);
+    mutable_column_device_view destination_col = destination_table.column(i);
+
+    if (source_col.has_nulls()) {
+      size_type destination_row_base = blockIdx.x * blockDim.x;
+
+      while (destination_row_base < destination_table.num_rows()) {
+	size_type destination_row = destination_row_base + threadIdx.x;
```

Comment:
I realize this was in the original code, but could you remove the mixture of tabs and spaces? Makes it hard to review in GitHub...